### ISSUE TITLE
Cannot add WebAuthn security keys without adding TOTP first

### ISSUE THREAD

**Issue by trwnh** [mastodon/mastodon#16693]
### Expected behaviour

I should be able to add WebAuthn security keys first/alone if I wish

### Actual behaviour

I must add TOTP before I can add WebAuth security keys

### Steps to reproduce the problem

1. Go to Settings > Accounts > Two-factor Auth
2. No option to add Security keys until you add Authenticator app

### Specifications

mastodon.social

**Comment by Be-ing** [comment]
To elaborate why this is a problem, TOTP requires users to manually input secrets. This makes it more cumbersome to use and vulnerable to phishing, whereas FIDO2 does not have those problems.

**Comment by LeeteqXV** [comment]
Just for the clarity: WebAuthn, also referred to as "paswordLESS", is actually replacing the 1st factor - namely the password - not the second factor. It is not a 2FA alternative.

Policywise, any site should have the option to choose whether or not to use 2FA, but still be able to choose to offer WebAuthn instead of passwords, and some sites might even want to say to all users that do NOT want to use 2FA, that then they must use WebAuthn instead of "just password", because passwords alone is worse...
(allthough that is not very good, as someone who gets hold of the usb key can then get in without any other information, so it is always recommended to "enforce" 2FA anyway, IMO)

- regardless, this issue should move ahead anyway.

**Comment by LeeteqXV** [comment]
One bit further on that last comment:

1. With "just password", who you are at risk to, are hackers on remote devices who may hack into the account, especially if the user has been using the same password on multiple services, and at least one of them has had a password leak/hack.

2. With "just WebAuthn", the situation is the opposite; no remote hackers can hack into your account, but anyone who are physically sufficiently near you to get hold of your usb key, can get into your account.

3. The person who want to (mis-)use your WebAuthn key, ALSO need to know both the URL to the services where you use it, AND your user name or email used there, or it will not be possible to get into your account. The key itself is not enough.

**Comment by Be-ing** [comment]
I agree that using WebAuthn for one factor authentication would be a nice option. That could also allow creating accounts that are not associated with email addresses.

> allthough that is not very good, as someone who gets hold of the usb key can then get in without any other information, so it is always recommended to "enforce" 2FA anyway, IMO

If this is a concern, FIDO2 can optionally use PINs and/or biometrics. WebAuthn can require this by requiring user verification in addition to user presence.

**Comment by santiagorodriguez96** [comment]
Hello guys!

The idea behind enforcing TOTP to be enabled before being able to enable WebAuthn is that you can have a fallback 2FA option in case you don't have your security key with you or even if you have lost it. In those cases, you'll still be able to log in via using TOTP.

After you have enabled WebAuthn as a 2FA, you don't need to use TOTP codes in order to log in if you have your security key with you.

I agree with you guys that WebAuthn is preferred in order to log in for all the reasons you said, but I do think it's wise to have a fallback option in order to being able to log in, thus enforcing TOTP seems like a healthy constraint to me.

What do you guys think?

**Comment by Be-ing** [comment]
You're making a big assumption that a TOTP code is harder to lose than a FIDO authenticator. 

> I do think it's wise to have a fallback option in order to being able to log in

This solution to this is to *require* registering a minimum of 2 additional factors (any combination of FIDO tokens and TOTP codes) and strongly encourage registering >2 to enable 2FA.

**Comment by BirdInFire** [comment]
> Hello guys!
> 
> The idea behind enforcing TOTP to be enabled before being able to enable WebAuthn is that you can have a fallback 2FA option in case you don't have your security key with you or even if you have lost it. In those cases, you'll still be able to log in via using TOTP.
> 
> After you have enabled WebAuthn as a 2FA, you don't need to use TOTP codes in order to log in if you have your security key with you.
> 
> I do agree with you that WebAuthn is preferred in order to log in for all the reasons you said, but I do think it's wise to have a fallback option in order to being able to log in, thus enforcing TOTP seems like a healthy constraint to me.
> 
> What do you guys think?

And what about user who have 2 security key and do not want totp at all ?
I'm for letting the choice to the user to select the security he want (only a warning when less than 2 is enabled).

**Comment by e-lisa** [comment]
> After you have enabled WebAuthn as a 2FA, you don't need to use TOTP codes in order to log in if you have your security key with you.
> 
> I do agree with you that WebAuthn is preferred in order to log in for all the reasons you said, but I do think it's wise to have a fallback option in order to being able to log in, thus enforcing TOTP seems like a healthy constraint to me.
> 
> What do you guys think?

This seems like a very dangerous way to be handling authentication, trying to re-invent the wheel. The general way this is handled is to generate a series of SINGLE USE backup code that can be used to bypass 2FA. Here is why using "TOTP" as a backup code replacement/requirement would be a bad idea:

1. Some people use the same device for TOTP and WebAuthn, thus if they loose one they loose the other
2. Some people do not have their device configured to generate a TOTP code at all, and find themselves unable to set up 2Fa at all
3. Gatekeeping 2FA with another 2FA feels like just asking the user to not bother (what if they only have WebAuthn set up, the average user is not going to set up another system for a one time thing they'll never use again)

I think security should be easy, or it wont be used at all. Let the user pick what 2FA option they want, and offer a single use backup code system for when they loose their device. The user should be told at 2FA creation that they cannot recover their backup codes, so they must store them in a safe place (encrypted password database, on paper, etc). I was under the impression that backup codes were already implemented in Mastodon via #563. Did something change?

IMHO: I'd like to set up 2FA on Mastodon, but having the feature gatekeeped behind TOTP makes me skeptical.

**Comment by GuillaumeRoss** [comment]
With mobile devices supporting FIDO2, I wish Mastodon would:

1) Support multiple security keys (pre-requisite to not requiring TOTP at all as a backup method)
2) Allowed setting them up without doing TOTP first
3) Allowed using Passkeys as a single factor for people who feel like that is secure enough.

**Comment by debbiegoldsmith** [comment]
Ideally, I'd like to be able to add WebAuthn to an existing account, and then turn off password authentication altogether, and then optionally add additional protection like a second WebAuthn factor. It shouldn't be just for new accounts.

**Comment by e-lisa** [comment]
> > @BirdInFire Agreed that passwordless with Passkeys is safer than 1FA, however, 2FA with security keys only does provide a small amount of additional security in the case of the loss of a device. For most people and most cases if Mastodon supported Passkeys as single factor it would be the right decision.
> 
> to fix that all mastodon have to do is to (when only use passwordless) require 2 yubikey or a yubikey and a passkey etc. If you lose one the second is there.

IMHO this is a fairly reasonable take

> Ideally, I'd like to be able to add WebAuthn to an existing account, and then turn off password authentication altogether, and then optionally add additional protection like a second WebAuthn factor. It shouldn't be just for new accounts.

This feels outside of the scope of this issue. As it seems to be approaching a new feature, and not addressing the current issue of WebAuthn being gated.

**Comment by bdewater** [comment]
> to fix that all mastodon have to do is to (when only use passwordless) require 2 yubikey or a yubikey and a passkey etc. If you lose one the second is there.

It seems to me this what the recovery codes are for, setting up TOTP twice is not a requirement for example? Which is also why I am not really following the original idea of having to set up TOTP before WebAuthn can be enabled.

The [goal of multi-device FIDO credentials (aka 'passkeys')](https://media.fidoalliance.org/wp-content/uploads/2022/03/How-FIDO-Addresses-a-Full-Range-of-Use-Cases-March24.pdf#page=5) is to not need these workarounds anymore (i.e. "survive device loss"). [Apple](https://support.apple.com/en-ca/HT213305) and [Google](https://security.googleblog.com/2022/10/SecurityofPasskeysintheGooglePasswordManager.html) are the first ones to implement this but [there are signals](https://github.com/passkeydeveloper/discussions/discussions/1) password managers like 1Password and Dashlane will follow in 2023.

**Comment by santiagorodriguez96** [comment]
Hey guys! I've been reading the different comments and thoughts shared here and I'm also starting to think it makes sense to remove the TOTP requirement – specially after, as @bdewater was mentioning, the (incoming) introduction of multi-device FIDO credentials 🙂 

I'll try to open a PR for removing it on the next couple of days 👀

**Comment by reynir** [comment]
I would very much like to see the TOTP requirement lifted. I have >5 U2F devices, and I have not enabled 2FA on any of my mastodon accounts yet because I don't have or want to use TOTP software.

I am sad and annoyed the thread has been derailed about password-less talk which this issue clearly was not about. I look forward to your PR, @santiagorodriguez96 !

**Comment by fdellwing** [comment]
Just a quick question for all of you pushing this issue:

How did you setup MFA in GitHub without first using TOTP? Isn't that industry standard to first require a TOTP device before allowing FIDO2?

**Comment by reynir** [comment]
It's a really misguided industry standard. It might have had some merit back when WebAuthn/U2F support in browsers weren't as pervasive, and you couldn't be sure you could use your security key on a new device.

**Comment by e-lisa** [comment]
> Just a quick question for all of you pushing this issue:
> 
> How did you setup MFA in GitHub without first using TOTP? Isn't that industry standard to first require a TOTP device before allowing FIDO2?

1. What industry "standard" could you please list a the specific "standard" you are talking about?
2. I just added my FIDO2 device, no TOTP required on GitHub. Unless they changed things, you can still add your security key with out TOTP of any kind.

**Comment by fdellwing** [comment]
> What industry "standard" could you please list a the specific "standard" you are talking about?

I can not because there is not written standard, but a lot of apps (I use) do it that way. That is why there is a question mark at the end.

> I just added my FIDO2 device, no TOTP required on GitHub. Unless they changed things, you can still add your security key with out TOTP of any kind.

How? I tried it yesterday with a fresh account and the only options are TOTP and SMS, other options are only available afterwards.

![grafik](https://user-images.githubusercontent.com/22813377/204091052-152ba522-a315-4f9f-ac97-8de5afe36ed0.png)
![grafik](https://user-images.githubusercontent.com/22813377/204091074-90aa90f1-9c86-489d-8eeb-845cdf73a514.png)

**Comment by e-lisa** [comment]
>> What industry "standard" could you please list a the specific "standard" you are talking about?

> I can not because there is not written standard, but a lot of apps (I use) do it that way. That is why there is a question mark at the end.

I think its time for people to stop hijacking this thread. Unless you can provide an actual standard or technical reason why this cannot move forward, I think its time to stop trying to argue against this feature request. Using "unwritten standards" is a very bad way to move forward. The FIDO2 standard itself does not require nor recommend "use of TOTP before enrolling." Full stop.

Personally I think requiring people to use an app (TOTP) that they may not wish to install or use, is bad plan. Less people will use 2FA, some have even mentioned this on the thread.

**Comment by mistial-dev** [comment]
> Just for the clarity: WebAuthn, also referred to as "passwordLESS", is actually replacing the 1st factor - namely the password - not the second factor. It is not a 2FA alternative.

Microsoft is requiring tokens have a PIN set for passwordless.  The possession of the token then is the first factor, and the possession of the PIN the second.

By writing the account name to the token, the user inserts and pins (like a smart card), achieving convenient two factor auth.

**Comment by debbiegoldsmith** [comment]
Right, there is no need for a second factor with WebAuthn unless the user desires it. I don't know why Microsoft requires a PIN. Apple's approach is biometric and doesn't require a PIN (or even allow for one).

**Comment by mistial-dev** [comment]
They require a PIN for tokens like yubikeys, because otherwise it would be one factor.On Sun, Dec 18, 2022 at 1:56 PM, debbiegoldsmith ***@***.***> wrote:  
Right, there is no need for a second factor with WebAuthn unless the user desires it. I don't know why Microsoft requires a PIN. Apple's approach is biometric and doesn't require a PIN (or even allow for one).

**Comment by debbiegoldsmith** [comment]
It's fine if 2FA is *allowed* with WebAuthn, but it shouldn't be *required*. As I mentioned, some approaches to WebAuthn don't require (and would not benefit from) 2FA. Any approach using biometrics, for example, already has the same level of security (possession of a physical object, plus the biometrics). Adding a PIN would make it 3FA.

**Comment by santiagorodriguez96** [comment]
> Any progress on this ? because it become quite annoying in 2023 to have this OTP requirement !

Hey! I've started working on it. Sorry for the delay 🙂

**Comment by schwaboy** [comment]
> The idea behind enforcing TOTP to be enabled before being able to enable WebAuthn is that you can have a fallback 2FA option in case you don't have your security key with you or even if you have lost it. In those cases, you'll still be able to log in via using TOTP.

Isn't this what Recovery Codes are for?

**Comment by cmeyer** [comment]
> > The idea behind enforcing TOTP to be enabled before being able to enable WebAuthn is that you can have a fallback 2FA option in case you don't have your security key with you or even if you have lost it. In those cases, you'll still be able to log in via using TOTP.
> 
> Isn't this what Recovery Codes are for?

Why can't WebAuthn be treated as one of several possibilities for authentication? I like to use WebAuthn for the convenience of not having to type my password. But I'm also happy saving my password as the backup mechanism in case I lose access to the machine on which WebAuthn is enabled. For that matter, one should be able to enable WebAuthn on multiple devices (multiple computers, a phone, a tablet, etc.).

**Comment by hex-m** [comment]
@cmeyer Webauthn is currently used as a second factor. https://github.com/mastodon/mastodon/issues/18341 is where passwordless is discussed. When you think about syncing Webauthn credentials to multiple devices, look into https://github.com/mastodon/mastodon/issues/20383 - otherwise you can register every device separately.

**Comment by BirdInFire** [comment]
Any ETA change ?

**Comment by aaronmdjones** [comment]
This is still annoying me even today. Fortunately, you can set up TOTP by generating the OTP with a command-line tool and then throw the seed away instead of saving it. This is what I've done.

It does still leave GitHub offering to allow you to sign in via TOTP instead (on the security key screen) though, which is annoying as it gives the impression that that can work. Oh well.
